File: 03_Sorting Techniques/04_SelectionSortRecursive/04_SelectionSortRecursive.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;


int minIndex(vector<int> nums, int i, int j)
{   
    if (i == j)
        return i;
 
    int k = minIndex(nums, i + 1, j);
 
    return (nums[i] < nums[k])? i : k;
}

void selectionSortRecursive(vector<int> &nums, int index = 0) {
    if (index == nums.size() - 1)
        return;
    
    int k = minIndex(nums, index, nums.size() - 1);

    if (k != index) 
        swap(nums[k], nums[index]);
    
    selectionSortRecursive(nums, index + 1);
    
}


int main() {
    vector<int> nums = {100, -22, 20, 11, 4, 2, 200, 24};
    selectionSortRecursive(nums);
    for (auto x: nums) {
        cout << x << ' ';
    }
    return 0;
}

File: 03_Sorting Techniques/04_SelectionSortRecursive/04_SelectionSortRecursive.js

function selectMin(nums, start, end = nums.length - 1, min_idx = start) {
    if (start >= end) {
        return min_idx;
    }
    
    if (nums[start + 1] < nums[min_idx]) {
        min_idx = start + 1;
    }
    
    return selectMin(nums, start + 1, end, min_idx);
}

function selectionSortRecursive(nums, start = 0, end = nums.length - 1) {
    
    if (start >= end) {
        return nums;
    }

    let min_idx =  selectMin(nums, start);
    if (nums[start] != nums[min_idx]) {
        [nums[start], nums[min_idx]] = [nums[min_idx], nums[start]];
    }
    
    return selectionSortRecursive(nums, start + 1, end)
}


console.log(selectionSortRecursive([100, 99, 98,97,-1]));